# Practical Application III: Comparing Classifiers

**Overview**: In this practical application, your goal is to compare the performance of the classifiers we encountered in this section, namely K Nearest Neighbor, Logistic Regression, Decision Trees, and Support Vector Machines.  We will utilize a dataset related to marketing bank products over the telephone.  



### Getting Started

Our dataset comes from the UCI Machine Learning repository [link](https://archive.ics.uci.edu/ml/datasets/bank+marketing).  The data is from a Portugese banking institution and is a collection of the results of multiple marketing campaigns.  We will make use of the article accompanying the dataset [here](CRISP-DM-BANK.pdf) for more information on the data and features.



### Problem 1: Understanding the Data

To gain a better understanding of the data, please read the information provided in the UCI link above, and examine the **Materials and Methods** section of the paper.  How many marketing campaigns does this data represent?

The data was gathered from 17 direct marketing campaigns that took place between May 2008 and November 2010, amounting to 79,354 total contacts.

### Problem 2: Read in the Data

Use pandas to read in the dataset `bank-additional-full.csv` and assign to a meaningful variable name.

In [31]:
import pandas as pd

In [33]:
df = pd.read_csv('data/bank-additional-full.csv', sep = ';')

In [17]:
df.head()

age        job  marital    education  default housing loan    contact  \
0   56  housemaid  married     basic.4y       no      no   no  telephone   
1   57   services  married  high.school  unknown      no   no  telephone   
2   37   services  married  high.school       no     yes   no  telephone   
3   40     admin.  married     basic.6y       no      no   no  telephone   
4   56   services  married  high.school       no      no  yes  telephone   

  month day_of_week  ...  campaign  pdays  previous     poutcome emp.var.rate  \
0   may         mon  ...         1    999         0  nonexistent          1.1   
1   may         mon  ...         1    999         0  nonexistent          1.1   
2   may         mon  ...         1    999         0  nonexistent          1.1   
3   may         mon  ...         1    999         0  nonexistent          1.1   
4   may         mon  ...         1    999         0  nonexistent          1.1   

   cons.price.idx  cons.conf.idx  euribor3m  nr.employed   y  
0          93.994          -36.4      4.857       5191.0  no  
1          93.994          -36.4      4.857       5191.0  no  
2          93.994          -36.4      4.857       5191.0  no  
3          93.994          -36.4      4.857       5191.0  no  
4          93.994          -36.4      4.857       5191.0  no  

[5 rows x 21 columns]

### Problem 3: Understanding the Features


Examine the data description below, and determine if any of the features are missing values or need to be coerced to a different data type.


```
Input variables:
# bank client data:
1 - age (numeric)
2 - job : type of job (categorical: 'admin.','blue-collar','entrepreneur','housemaid','management','retired','self-employed','services','student','technician','unemployed','unknown')
3 - marital : marital status (categorical: 'divorced','married','single','unknown'; note: 'divorced' means divorced or widowed)
4 - education (categorical: 'basic.4y','basic.6y','basic.9y','high.school','illiterate','professional.course','university.degree','unknown')
5 - default: has credit in default? (categorical: 'no','yes','unknown')
6 - housing: has housing loan? (categorical: 'no','yes','unknown')
7 - loan: has personal loan? (categorical: 'no','yes','unknown')
# related with the last contact of the current campaign:
8 - contact: contact communication type (categorical: 'cellular','telephone')
9 - month: last contact month of year (categorical: 'jan', 'feb', 'mar', ..., 'nov', 'dec')
10 - day_of_week: last contact day of the week (categorical: 'mon','tue','wed','thu','fri')
11 - duration: last contact duration, in seconds (numeric). Important note: this attribute highly affects the output target (e.g., if duration=0 then y='no'). Yet, the duration is not known before a call is performed. Also, after the end of the call y is obviously known. Thus, this input should only be included for benchmark purposes and should be discarded if the intention is to have a realistic predictive model.
# other attributes:
12 - campaign: number of contacts performed during this campaign and for this client (numeric, includes last contact)
13 - pdays: number of days that passed by after the client was last contacted from a previous campaign (numeric; 999 means client was not previously contacted)
14 - previous: number of contacts performed before this campaign and for this client (numeric)
15 - poutcome: outcome of the previous marketing campaign (categorical: 'failure','nonexistent','success')
# social and economic context attributes
16 - emp.var.rate: employment variation rate - quarterly indicator (numeric)
17 - cons.price.idx: consumer price index - monthly indicator (numeric)
18 - cons.conf.idx: consumer confidence index - monthly indicator (numeric)
19 - euribor3m: euribor 3 month rate - daily indicator (numeric)
20 - nr.employed: number of employees - quarterly indicator (numeric)

Output variable (desired target):
21 - y - has the client subscribed a term deposit? (binary: 'yes','no')
```



The following features contain missing data (value 'unknown') : job, marital, education, default, housing, loan
The feature pdays uses 999 as a placeholder and needs to be converted into a binary.
The output variable y has yes/no and needs to be converted into 1/0.
The features contact, month, day_of_ are categorical and can be one hot encoded.
The feature duration can be discarded for a realistic model.
The feature poutcome contains 'nonexistent' which can be one hot encoded or replaced with 'failure'.

We may not use all the features to predict the outcome.
We can select the set of features that are more likely to influence the outcome e.g. 'age', 'job', 'marital', 'education', 'default', 'housing', 'loan' etc. We don't need to consider the consumer price index, employment variation rate etc. for modeling this data.

### Problem 4: Understanding the Task

After examining the description and data, your goal now is to clearly state the *Business Objective* of the task.  State the objective below.

In [20]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 41188 entries, 0 to 41187
Data columns (total 21 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   age             41188 non-null  int64  
 1   job             41188 non-null  object 
 2   marital         41188 non-null  object 
 3   education       41188 non-null  object 
 4   default         41188 non-null  object 
 5   housing         41188 non-null  object 
 6   loan            41188 non-null  object 
 7   contact         41188 non-null  object 
 8   month           41188 non-null  object 
 9   day_of_week     41188 non-null  object 
 10  duration        41188 non-null  int64  
 11  campaign        41188 non-null  int64  
 12  pdays           41188 non-null  int64  
 13  previous        41188 non-null  int64  
 14  poutcome        41188 non-null  object 
 15  emp.var.rate    41188 non-null  float64
 16  cons.price.idx  41188 non-null  float64
 17  cons.conf.idx   41188 non-null 

The business objective is to create a predictive model that identifies the key characteristics of clients who are likely to say "yes" to a long-term bank deposit. With this model, the bank can more accurately target a smaller, higher-quality group of potential customers.

### Problem 5: Engineering Features

Now that you understand your business objective, we will build a basic model to get started.  Before we can do this, we must work to encode the data.  Using just the bank information features, prepare the features and target column for modeling with appropriate encoding and transformations.

In [35]:
# Define the specific bank client features and the target variable to be used
bank_client_features = ['age', 'job', 'marital', 'education', 'default', 'housing', 'loan']
target = 'y'

# Create a new DataFrame with only the selected columns to avoid modifying the original data
df_model_data = df[bank_client_features + [target]].copy()

# Encode the binary target variable 'y'
df_model_data[target] = df_model_data[target].map({'yes': 1, 'no': 0})

# Identify the categorical features that need to be converted to a numeric format
categorical_features = ['job', 'marital', 'education', 'default', 'housing', 'loan']

# Apply one-hot encoding to the categorical features
df_encoded = pd.get_dummies(df_model_data, columns=categorical_features, drop_first=True)

print("\nFirst 5 rows of the prepared data:")
print(df_encoded.head())

print("\nInfo of the prepared data:")
df_encoded.info()


First 5 rows of the prepared data:
   age  y  job_blue-collar  job_entrepreneur  job_housemaid  job_management  \
0   56  0            False             False           True           False   
1   57  0            False             False          False           False   
2   37  0            False             False          False           False   
3   40  0            False             False          False           False   
4   56  0            False             False          False           False   

   job_retired  job_self-employed  job_services  job_student  ...  \
0        False              False         False        False  ...   
1        False              False          True        False  ...   
2        False              False          True        False  ...   
3        False              False         False        False  ...   
4        False              False          True        False  ...   

   education_illiterate  education_professional.course  \
0               

In [27]:
df.head()

age        job  marital    education default housing loan    contact month  \
0   56  housemaid  married     basic.4y      no      no   no  telephone   may   
1   57   services  married  high.school    <NA>      no   no  telephone   may   
2   37   services  married  high.school      no     yes   no  telephone   may   
3   40     admin.  married     basic.6y      no      no   no  telephone   may   
4   56   services  married  high.school      no      no  yes  telephone   may   

  day_of_week  campaign  pdays  previous     poutcome  emp.var.rate  \
0         mon         1    999         0  nonexistent           1.1   
1         mon         1    999         0  nonexistent           1.1   
2         mon         1    999         0  nonexistent           1.1   
3         mon         1    999         0  nonexistent           1.1   
4         mon         1    999         0  nonexistent           1.1   

   cons.price.idx  cons.conf.idx  euribor3m  nr.employed  y  
0          93.994          -36.4      4.857       5191.0  0  
1          93.994          -36.4      4.857       5191.0  0  
2          93.994          -36.4      4.857       5191.0  0  
3          93.994          -36.4      4.857       5191.0  0  
4          93.994          -36.4      4.857       5191.0  0

### Problem 6: Train/Test Split

With your data prepared, split it into a train and test set.

In [39]:
from sklearn.model_selection import train_test_split

X = df_encoded.drop('y', axis=1)
y = df_encoded['y']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y)

### Problem 7: A Baseline Model
Before we build our first model, we want to establish a baseline.  What is the baseline performance that our classifier should aim to beat?

In [43]:
from sklearn.dummy import DummyClassifier

dummy_clf = DummyClassifier().fit(X_train, y_train)
baseline_score = dummy_clf.score(X_test, y_test)
print(baseline_score)

0.8873512988589464


### Problem 8: A Simple Model

Use Logistic Regression to build a basic model on your data.  

In [47]:
from sklearn.linear_model import LogisticRegression

lreg = LogisticRegression(random_state=42, class_weight='balanced', max_iter=1000)
lreg.fit(X_train, y_train)

LogisticRegression(class_weight='balanced', max_iter=1000, random_state=42)

### Problem 9: Score the Model

What is the accuracy of your model?

In [51]:
from sklearn.metrics import accuracy_score

y_pred = lreg.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)
print(accuracy)

0.5830298616168973


### Problem 10: Model Comparisons

Now, we aim to compare the performance of the Logistic Regression model to our KNN algorithm, Decision Tree, and SVM models.  Using the default settings for each of the models, fit and score each.  Also, be sure to compare the fit time of each of the models.  Present your findings in a `DataFrame` similar to that below:

| Model | Train Time | Train Accuracy | Test Accuracy |
| ----- | ---------- | -------------  | -----------   |
|     |    |.     |.     |

In [66]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.svm import SVC
import time

In [73]:
models = {
    "Logistic Regression": LogisticRegression(random_state=42, class_weight='balanced', max_iter=1000),
    "KNN": KNeighborsClassifier(),
    "Decision Tree": DecisionTreeClassifier(random_state=42),
    "SVM": SVC(random_state=42)
}

results = []

for model_name, model in models.items():
    start_time = time.time()
    model.fit(X_train, y_train)
    end_time = time.time()
    train_time = end_time - start_time

    train_accuracy = model.score(X_train, y_train)
    test_accuracy = model.score(X_test, y_test)

    # Store the results
    results.append({
        "Model": model_name,
        "Train Time": round(train_time, 4),
        "Train Accuracy": round(train_accuracy, 4),
        "Test Accuracy": round(test_accuracy, 4)
    })

results_df = pd.DataFrame(results)
print(results_df)

                 Model  Train Time  Train Accuracy  Test Accuracy
0  Logistic Regression      0.9218          0.5895         0.5830
1                  KNN      0.0184          0.8914         0.8775
2        Decision Tree      0.1379          0.9171         0.8640
3                  SVM     15.9440          0.8873         0.8874


Here we see that SVM provides the best Test Accuracy and Decision Tree provides the best Train accuracy.

### Problem 11: Improving the Model

Now that we have some basic models on the board, we want to try to improve these.  Below, we list a few things to explore in this pursuit.

- More feature engineering and exploration.  For example, should we keep the gender feature?  Why or why not?
- Hyperparameter tuning and grid search.  All of our models have additional hyperparameters to tune and explore.  For example the number of neighbors in KNN or the maximum depth of a Decision Tree.  
- Adjust your performance metric

- Gender is not an useful feature for this prediction as per the corresponding research paper. Success rate was almost equal for males and females. Hence it maybe considered as noise.

- The following parameters could be changed :
  - number of neighbors in KNN
  - max depth or min samples leaf in the Decision Tree
  - max iter in Logistic Regression
  - penalty term or kernel type in SVM

A common technique for this is Grid Search, which systematically works through all the hyperparameter combinations you specify and finds the set that yields the best performance.

As the data contains a skewed number of negatives and positives the accuracy can be misleading.
Better metrics could be:
- Recall
- Precision
- F1

In [ ]:
from sklearn.model_selection import train_test_split, GridSearchCV

# KNN params
param_grid_knn = {
    'n_neighbors': [3, 5, 7, 9, 11],
    'weights': ['uniform', 'distance'],
    'metric': ['euclidean', 'manhattan']
}

# Decision Tree params
param_grid_dt = {
    'criterion': ['gini', 'entropy'],
    'max_depth': [None, 10, 20, 30],
    'min_samples_leaf': [1, 5, 10]
}

# SVM params
param_grid_svm = {
    'C': [0.1, 1, 10],
    'kernel': ['linear', 'rbf']
}

grid_knn = GridSearchCV(KNeighborsClassifier(), param_grid_knn, cv=3, scoring='f1', n_jobs=-1, verbose=1)
grid_knn.fit(X_train, y_train)
print("\nBest Parameters for KNN:", grid_knn.best_params_)
print("Best F1-Score for KNN:", grid_knn.best_score_)

grid_dt = GridSearchCV(DecisionTreeClassifier(random_state=42), param_grid_dt, cv=3, scoring='f1', n_jobs=-1, verbose=1)
grid_dt.fit(X_train, y_train)
print("\nBest Parameters for Decision Tree:", grid_dt.best_params_)
print("Best F1-Score for Decision Tree:", grid_dt.best_score_)

grid_svm = GridSearchCV(SVC(random_state=42), param_grid_svm, cv=3, scoring='f1', n_jobs=-1, verbose=1)
grid_svm.fit(X_train, y_train)
print("\nBest Parameters for SVM:", grid_svm.best_params_)
print("Best F1-Score for SVM:", grid_svm.best_score_)

Fitting 3 folds for each of 20 candidates, totalling 60 fits


C:\Users\Owner\anaconda3\Lib\site-packages\sklearn\model_selection\_search.py:1102: UserWarning: One or more of the test scores are non-finite: [nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan
 nan nan]
  warnings.warn(



Best Parameters for KNN: {'metric': 'euclidean', 'n_neighbors': 3, 'weights': 'uniform'}
Best F1-Score for KNN: nan
Fitting 3 folds for each of 24 candidates, totalling 72 fits

Best Parameters for Decision Tree: {'criterion': 'entropy', 'max_depth': None, 'min_samples_leaf': 1}
Best F1-Score for Decision Tree: 0.139711713463451
Fitting 3 folds for each of 6 candidates, totalling 18 fits


##### Questions